trace-utils/src/span/mod.rs

```diff
@@ -124,8 +129,127 @@ where
 }
 
+/// The generic representation of a V04 span event.
+/// `T` is the type used to represent strings in the span event.
+#[derive(Clone, Debug, Default, PartialEq, Serialize)]
+pub struct SpanEvent<T>
+where
+    T: SpanText,
+{
+    pub time_unix_nano: u64,
+    pub name: T,
+    #[serde(skip_serializing_if = "HashMap::is_empty")]
+    pub attributes: HashMap<T, AttributeAnyValue<T>>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum AttributeAnyValue<T>
+where
+    T: SpanText,
+{
+    SingleValue(AttributeArrayValue<T>),
+    Array(Vec<AttributeArrayValue<T>>),
+}
+
+impl<T> Serialize for AttributeAnyValue<T>
+where
+    T: SpanText,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut state = serializer.serialize_struct("AttributeAnyValue", 2)?;
+
+        match self {
+            AttributeAnyValue::SingleValue(attribute) => {
+                serialize_attribute_array::<S, T>(&mut state, attribute)?;
+            }
+            AttributeAnyValue::Array(value) => {
+                let value_type: u8 = self.into();
+                state.serialize_field("type", &value_type)?;
+                state.serialize_field("array_value", value)?;
+            }
+        }
+
+        state.end()
+    }
+}
+
+impl<T> From<&AttributeAnyValue<T>> for u8
+where
+    T: SpanText,
+{
+    fn from(attribute: &AttributeAnyValue<T>) -> u8 {
+        match attribute {
+            AttributeAnyValue::SingleValue(value) => value.into(),
+            AttributeAnyValue::Array(_) => 4,
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum AttributeArrayValue<T>
+where
+    T: SpanText,
+{
+    String(T),
+    Boolean(bool),
+    Integer(i64),
+    Double(f64),
+}
+
+impl<T> Serialize for AttributeArrayValue<T>
+where
+    T: SpanText,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut state = serializer.serialize_struct("AttributeArrayValue", 2)?;
+        serialize_attribute_array::<S, T>(&mut state, self)?;
+        state.end()
+    }
+}
+
+fn serialize_attribute_array<S, T>(
+    state: &mut S::SerializeStruct,
+    attribute: &AttributeArrayValue<T>,
+) -> Result<(), <S>::Error>
+where
+    T: SpanText,
+    S: serde::Serializer,
+{
+    let attribute_type: u8 = attribute.into();
+    state.serialize_field("type", &attribute_type)?;
+    match attribute {
+        AttributeArrayValue::String(value) => state.serialize_field("string_value", value),
+        AttributeArrayValue::Boolean(value) => state.serialize_field("bool_value", value),
+        AttributeArrayValue::Integer(value) => state.serialize_field("int_value", value),
+        AttributeArrayValue::Double(value) => state.serialize_field("double_value", value),
+    }
+}
+
+impl<T> From<&AttributeArrayValue<T>> for u8
+where
+    T: SpanText,
+{
+    fn from(attribute: &AttributeArrayValue<T>) -> u8 {
```

Comment:
I know it's probably covered in your other tests, but it's not a bad idea to add a unit test specifically for this trait impl. It may seem unnecessary for such trivial logic, but could come in handy as part of future refactors.